### PR TITLE
MAINT: Move dtype string functions to python

### PR DESCRIPTION
**Pull request by eric-wieser**
Doing proper error checking of string concatenation in python super tedious.
As a results, we're not doing it, making us prone to exceptions turning into segfaults.
This moves all of the `__str__` and `__repr__ `functions, where speed is irrelevant, into np.core._dtype.

---

Minimal refactors made during the conversion from C - mainly just using `str.join` instead of manual loops. I'm sure more could be done here, but now that's it's python it should be easier for new contributors to modify.

CircleCI failures are because this is working off an earlier branch point

### CODE REVIEW

File: numpy/core/_dtype.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    else:
+        raise RuntimeError(
+            "Internal error: NumPy dtype unrecognized type number")
+
+
+def _byte_order_str(dtype):
+    """ Normalize byteorder to '<' or '>' """
+    # hack to obtain the native and swapped byte order characters
+    swapped = np.dtype(int).newbyteorder('s')
+    native = swapped.newbyteorder('s')
```
Comment (eric-wieser) [inline numpy/core/_dtype.py]: Is there a better way to get hold of these?

File: numpy/core/_dtype.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    """
+    Checks whether the structured data type in 'dtype'
+    has a simple layout, where all the fields are in order,
+    and follow each other with no alignment padding.
+
+    When this returns true, the dtype can be reconstructed
+    from a list of the field names and dtypes with no additional
+    dtype parameters.
+
+    Duplicates the C `is_dtype_struct_simple_unaligned_layout` functio.
```
Comment (eric-wieser) [inline numpy/core/_dtype.py]: I think we should add this as `dtype.ispacked` in another PR

File: numpy/core/src/multiarray/descriptor.c [hunk 3]
```diff
@@ -3167,462 +3167,36 @@ is_dtype_struct_simple_unaligned_layout(PyArray_Descr *dtype)
...
-            }
-            title = Py_None;
-            if (!PyArg_ParseTuple(tup, "Oi|O", &fld_dtype,
-                                            &fld_offset, &title)) {
-                PyErr_Clear();
-                return 0;
-            }
-            PyUString_ConcatAndDel(&ret, PyObject_Repr(title));
-            if (i != names_size - 1) {
-                PyUString_ConcatAndDel(&ret, PyUString_FromString(","));
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/descriptor.c]: This kind of thing segfaults if `PyUString_FromString(...)` returns `NULL` (ie, we run out of memory). This type of thing what motivated the switch to python - getting this right everywhere is too hard and too verbose in C, and for no gain.

**Comment by eric-wieser** [comment]
Any thoughts on this?

**Comment by mattip** [comment]
Could you rebase this after the coverage integration? I would like to see that we are checking all the new code in tests.

**Comment by eric-wieser** [comment]
Good idea - done

File: numpy/core/_dtype.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+from __future__ import division, absolute_import, print_function
+
+import numpy as np
+
+
+def __str__(dtype):
+    if dtype.fields is not None:
+        return _struct_str(dtype, include_align=True)
+    elif dtype.subdtype:
+        return _subarray_str(dtype)
```
Comment (mattip) [inline numpy/core/_dtype.py]: This line not hit in tests, see [coverage](https://codecov.io/gh/numpy/numpy/commit/73e328c9c9c4729c28f0f2dec407aed168b5fbdf#D3-15)

File: numpy/core/_dtype.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    """ Normalize byteorder to '<' or '>' """
+    # hack to obtain the native and swapped byte order characters
+    swapped = np.dtype(int).newbyteorder('s')
+    native = swapped.newbyteorder('s')
+
+    byteorder = dtype.byteorder
+    if byteorder == '=':
+        return native.byteorder
+    if byteorder == 's':
+        return swapped.byteorder
```
Comment (mattip) [inline numpy/core/_dtype.py]: This line not hit in [tests](https://codecov.io/gh/numpy/numpy/commit/73e328c9c9c4729c28f0f2dec407aed168b5fbdf#D3-155)

**Comment by mattip** [comment]
Coverage seems pretty good, only two blocks (other than exceptions) not covered

File: numpy/core/_dtype.py (hunk 1 above)
Comment (mattip) [reply numpy/core/_dtype.py]: There is `np.little_endian`or `sys.byteorder`, but then you have to decide '<' or '>' yourself.

File: numpy/core/_dtype.py (hunk 5 above)
Comment (eric-wieser) [reply numpy/core/_dtype.py]: Not sure if it's possible to hit - but this same test existed in the old C code.

Comment (eric-wieser) [reply numpy/core/_dtype.py]: Added a TODO

File: numpy/core/_dtype.py (hunk 4 above)
Comment (eric-wieser) [reply numpy/core/_dtype.py]: Test added

**Comment by eric-wieser** [comment]
@mattip: Coverage improved, I think

**Review by mattip — APPROVED** [review]

**Comment by mattip** [comment]
PyPy will JIT after ~1000 calls, if someone is calling repr(dtype) that many times I thing something is probably wrong with their code.

**Comment by ahaldane** [comment]
This seems to have triggered a circular import problem when using the dtype-transer debug mode. (set `NPY_DT_DBG_TRACING` to 1 at the top of `dtype_transfer.c`).

The problem seems to be that debug tracing tries to print a dtype before the dtype object has been created, I'm still trying to see what happened. It involves `getlimits.py` which has been the source of circular import issues in the past.

```python
>>> import numpy as np
Calculating dtype transfer from  to 
Traceback (most recent call last):
  File "/home/allan/nmpy_dev/build/testenv/lib/python3.7/site-packages/numpy/core/_dtype.py", line 23, in __repr__
    arg_str = _construction_repr(dtype, include_align=False)
  File "/home/allan/nmpy_dev/build/testenv/lib/python3.7/site-packages/numpy/core/_dtype.py", line 77, in _construction_repr
    return _scalar_str(dtype, short=short)
  File "/home/allan/nmpy_dev/build/testenv/lib/python3.7/site-packages/numpy/core/_dtype.py", line 81, in _scalar_str
    byteorder = _byte_order_str(dtype)
  File "/home/allan/nmpy_dev/build/testenv/lib/python3.7/site-packages/numpy/core/_dtype.py", line 151, in _byte_order_str
    swapped = np.dtype(int).newbyteorder('s')
AttributeError: module 'numpy' has no attribute 'dtype'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<frozen importlib._bootstrap>", line 200, in _lock_unlock_module
  File "<frozen importlib._bootstrap>", line 163, in _get_module_lock
SystemError: <built-in function acquire_lock> returned a result with an error set

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/allan/nmpy_dev/build/testenv/lib/python3.7/site-packages/numpy/__init__.py", line 142, in <module>
    from . import core
  File "/home/allan/nmpy_dev/build/testenv/lib/python3.7/site-packages/numpy/core/__init__.py", line 72, in <module>
    from . import getlimits
  File "/home/allan/nmpy_dev/build/testenv/lib/python3.7/site-packages/numpy/core/getlimits.py", line 125, in <module>
    tiny=_f16(2 ** -14))
  File "/home/allan/nmpy_dev/build/testenv/lib/python3.7/site-packages/numpy/core/getlimits.py", line 75, in __init__
    self.epsilon = self.eps = float_to_float(kwargs.pop('eps'))
  File "/home/allan/nmpy_dev/build/testenv/lib/python3.7/site-packages/numpy/core/getlimits.py", line 70, in <lambda>
    float_to_float = lambda v : _fr1(float_conv(v))
  File "/home/allan/nmpy_dev/build/testenv/lib/python3.7/site-packages/numpy/core/getlimits.py", line 29, in _fr1
    a = a.copy()
SystemError: <method 'copy' of 'numpy.ndarray' objects> returned a result with an error set
```

**Comment by charris** [comment]
@ahaldane Could you open an issue for this?

File: numpy/core/_dtype.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+        else:
+            return "'S%d'" % dtype.itemsize
+
+    elif dtype.type == np.unicode_:
+        if _isunsized(dtype):
+            return "'%sU'" % byteorder
+        else:
+            return "'%sU%d'" % (byteorder, dtype.itemsize / 4)
+
+    elif dtype.type == np.void:
```
Comment (eric-wieser) [inline numpy/core/_dtype.py]: Previously this caught subclasses of void too. Do we actually want that behavior?

**Pull request merged**